types/src/proof/ensure_proof.rs

```diff
@@ -0,0 +1,34 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module declares the macro ensure_proof!,
+//! which checks a condition for a proof and throws a ProofError
+//! if the proof fails.
+//! Currently, this macro only supports SparseMerkleProofs
```

Comment:
How do you intend to use this with others?

Reply:
There are other types of proofs in `types/src/proof/definition.rs` (e.g. `AccumulatorRangeProof`, `AccumulatorProof`) that we can later expand support for. For this PR I just wanted to implement it with `SparseMerkleProofs` to get feedback and iterate.

Reply:
Do you plan to do

```
match $proof {
    SparseMerkleProof { .. } => {
        ...
    }
    AccumulatorProof { .. } => {
        ...
    }
}
```

or something else in the macro? Note that the macro expansion happens at compile time, so as an example,

```
ensure_proof!(self, a == b, "message");
```

becomes

```
match self {
    SparseMerkleProof { .. } => {
        if !(a == b) { return Err(...); }
    }
    AccumulatorProof { .. } => {
        if !(a == b) { return Err(...); }
    }    
}
```

at compile time, which doesn't work.

Reply:
I see. I just pushed an alternative approach where we use a `Proof` trait to identify proofs in `ensure_proof!`. This seems to fix the compile issue?